File: homework4/Satyam_Mishra/Q1_BuildATrie.java
package homework4.Satyam_Mishra;

public class Q1_BuildATrie {
    static class Trie {
        Trie[] children;
        boolean validWord;

        public Trie() {
            children = new Trie[26];
            validWord = false;
        }
    }

    public static void insert(String word, Trie curr) {
        int n = word.length();

        for(int i = 0; i < n; i++) {
            int ch = word.charAt(i) - 'a';

            if(curr.children[ch] == null) {
                curr.children[ch] = new Trie();
            }
            curr = curr.children[ch];
        }

        curr.validWord = true;
    }

    public static boolean isValidWord(String word, Trie curr) {
        int n = word.length();

        for(int i = 0; i < n; i++) {
            int ch = word.charAt(i) - 'a';

            if(curr.children[ch] == null) return false;

            curr = curr.children[ch];
        }

        return curr.validWord;
    }

    public static boolean remove(int depth, String word, Trie curr) {
        if(curr == null) {
            return false;
        }

        if(depth == word.length()) {
            curr.validWord = false;
            return isEmpty(curr);
        }

        int ch = word.charAt(depth) - 'a';
        boolean shouldDeleteChild = remove(depth + 1, word, curr.children[ch]);

        if(shouldDeleteChild) {
            curr.children[ch] = null;
        } else {
            return false;
        }

        return !curr.validWord && isEmpty(curr);
    }

    private static boolean isEmpty(Trie curr) {
        for(int i = 0; i < 26; i++) {
            if(curr.children[i] != null) return false;
        }

        return true;
    }
}

File: homework4/Aditya_Prasad/q1_BuildATrie.cpp
// Technique : Trie as a parameter
// insert() - T.C = O(L), M.C = O(L)
// isValidWord() - T.C = O(L), M.C = O(1)
// remove() - T.C = O(L), M.C = O(L)
// where L is the length of the word

#include <bits/stdc++.h>
using namespace std;

class Trie {
private:
    struct TrieNode {
        vector<TrieNode*> children;
        bool validWord;

        TrieNode() {
            children.resize(26);
            validWord = false;
            for (int i = 0; i < 26; i++) {
                children[i] = nullptr;
            }
        }
    };

    TrieNode* root;

public:
    Trie() {
        root = new TrieNode();
    }

    void insert(string word) {
        TrieNode* node = root;

        for (int i = 0; i < word.size(); i++) {
            int idx = word[i] - 'a';

            if (node->children[idx] == nullptr) {
                node->children[idx] = new TrieNode();
            }

            node = node->children[idx];
        }

        node->validWord = true;
    }

    bool isValidWord(string word) {
        TrieNode* node = root;

        for (int i = 0; i < word.size(); i++) {
            int idx = word[i] - 'a';

            if (node->children[idx] == nullptr) {
                return false;
            }

            node = node->children[idx];
        }

        return node->validWord;
    }

    bool remove(string word, int depth, TrieNode* node) {
        if (depth == word.size()) {
            if (node->validWord == false) {
                return false;
            }
            node->validWord = false;
            return isEmpty(node);
        }
        bool shouldDeleteChild =
            remove(word, depth + 1, node->children[word[depth] - 'a']);

        if (shouldDeleteChild == true) {
            delete node->children[word[depth] - 'a'];
            node->children[word[depth] - 'a'] = nullptr;
        }

        return (node->validWord == false) && isEmpty(node);
    }

    bool isEmpty(TrieNode* node) {
        for (int i = 0; i < 26; i++) {
            if (node->children[i] != nullptr) {
                return false;
            }
        }
        return true;
    }
};

// Time Taken : 40 min